### PR TITLE
[Pipeline Viz] Zoom Controls

### PR DESCRIPTION
**Pull request by emilyzhong**
# Zoom Controls
* [ + | - ] zoom controls (similar to controls on DD maps)
* Switched zoom/pan libraries
* Double click and scroll to zoom 
* Misc UI adjustments (entry edge for first stage hidden, node alignment code, etc)

image
ezgif-2-7f848e674565

# Notes
* Icons for plus/minus controls are pulled from DD maps control

# Tests
Go to `samples/[SAMPLE_ID]/stage_results` with pipeline viz enabled account

### CODE REVIEW

File: app/assets/src/components/views/PipelineViz.jsx [hunk 1]
```diff
@@ -27,18 +28,19 @@ class PipelineViz extends React.Component {
...
 
   componentDidMount() {
     this.drawGraphs();
   }
 
-  handleMouseWheelZoom = e => {
-    const { zoomChangeInterval } = this.props;
-    const zoomChange = (e.deltaY < 0 ? 1 : -1) * zoomChangeInterval;
-    this.setState({ zoom: this.state.zoom + zoomChange });
+  zoomIn = () => {
```
Comment (jshoe) [inline app/assets/src/components/views/PipelineViz.jsx]: these could just be `zoomIn = this.panZoomContainer.zoomIn;`

File: app/assets/src/components/ui/controls/PlusMinusControl.jsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import cx from "classnames";
+import PropTypes from "prop-types";
+import React from "react";
+
+import PlusControlIcon from "~/components/ui/icons/PlusControlIcon.jsx";
+import MinusControlIcon from "~/components/ui/icons/MinusControlIcon.jsx";
+import cs from "./plus_minus_control.scss";
+
+const PlusMinusControl = ({
+  onClickPlus,
```
Comment (jshoe) [inline app/assets/src/components/ui/controls/PlusMinusControl.jsx]: nit: we tend to follow noun-verb i.e. `onPlusClick` (https://jaketrent.com/post/naming-event-handlers-react/)

File: app/assets/src/components/ui/controls/PlusMinusControl.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  onClickMinus,
+  plusDisabled,
+  minusDisabled,
+  className,
+}) => {
+  return (
+    <div className={cx(className, cs.plusMinusContainer)}>
+      <button disabled={plusDisabled} onClick={onClickPlus}>
```
Comment (jshoe) [inline app/assets/src/components/ui/controls/PlusMinusControl.jsx]: also is this styling already changing when 'disabled'?

Comment (emilyzhong) [reply app/assets/src/components/ui/controls/PlusMinusControl.jsx]: I think the button on its own changes the cursor, but other than that nothing is changed in styling (not too sure what it should look like disabled either)

File: app/assets/src/components/views/PipelineViz.jsx (hunk 1 above)
Comment (emilyzhong) [reply app/assets/src/components/views/PipelineViz.jsx]: I tried that (I think? Unless I am doing something wrong syntactically) but it errors thinking that `this.panZoomContainer` is undefined

Comment (emilyzhong) [reply app/assets/src/components/views/PipelineViz.jsx]: I can modify the `on[Plus/Minus]Click` prop to `() => this.panZoomContainer.zoom[In/Out]` though

Comment (jshoe) [reply app/assets/src/components/views/PipelineViz.jsx]: oh it probably is undefined at that point. nvm then

File: app/assets/src/components/views/PipelineViz.jsx [hunk 4]
```diff
@@ -1,10 +1,11 @@
 import React from "react";
 import { mapValues } from "lodash/fp";
 import PropTypes from "prop-types";
-import ReactPanZoom from "@ajainarayanan/react-pan-zoom";
+import { PanZoom } from "react-easy-panzoom";
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: 👍

File: app/assets/src/components/views/PipelineViz.jsx (hunk 1 above)
Comment (tfrcarvalho) [reply app/assets/src/components/views/PipelineViz.jsx]: Why not just use this last suggestion directly . on the component below?

File: app/assets/src/components/views/PipelineViz.jsx [hunk 5]
```diff
@@ -125,7 +128,19 @@ class PipelineViz extends React.Component {
...
         } else {
           // Connect beginning steps to input node
-          intraEdgeData.push({ from: START_NODE_ID, to: i });
+          intraEdgeData.push({
+            from: START_NODE_ID,
+            to: i,
+            ...(index == 0
+              ? {
+                  color: {
+                    color: backgroundColor,
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: Is this background color for an edge? Why does it need background color?

File: app/assets/src/components/views/PipelineViz.jsx [hunk 6]
```diff
@@ -343,10 +349,23 @@ class PipelineViz extends React.Component {
...
+          zoomSpeed={3}
+          ref={ref => {
+            this.panZoomContainer = ref;
+          }}
+        >
           <div className={cs.pipelineViz}>{stageContainers}</div>
-        </ReactPanZoom>
+        </PanZoom>
+        <PlusMinusControl
+          onPlusClick={this.zoomIn}
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: Maybe user `(this.panZoomContainer || {}).zoomIn` here?

File: app/assets/src/components/views/PipelineViz.jsx [hunk 7]
```diff
@@ -343,10 +349,23 @@ class PipelineViz extends React.Component {
...
+          onMinusClick={this.zoomOut}
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: `(this.panZoomContainer || {}).zoomOut `

File: app/assets/src/components/views/PipelineViz.jsx [hunk 8]
```diff
@@ -343,10 +349,23 @@ class PipelineViz extends React.Component {
...
 
     return (
-      <div onWheel={this.handleMouseWheelZoom}>
-        <ReactPanZoom zoom={this.state.zoom}>
+      <div>
+        <PanZoom
+          className={cs.panZoomContainer}
+          minZoom={zoomMin}
+          maxZoom={zoomMax}
+          zoomSpeed={3}
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: Can we move to a prop with default 3?

File: app/assets/src/components/views/PipelineViz.jsx (hunk 5 above)
Comment (emilyzhong) [reply app/assets/src/components/views/PipelineViz.jsx]: In the mocks the first stage doesn't have an entry edge to the first step, but in order to center the graph vertically I need to connect into the starting node, so the first edge to the first step of the first stage has a background color the same as the background to hide it. 

I'll add a comment in the code that explains this!

File: app/assets/src/components/views/PipelineViz.jsx [hunk 9]
```diff
@@ -343,10 +342,27 @@ class PipelineViz extends React.Component {
...
+          ref={ref => {
+            this.panZoomContainer = ref;
+          }}
+        >
           <div className={cs.pipelineViz}>{stageContainers}</div>
-        </ReactPanZoom>
+        </PanZoom>
+        <PlusMinusControl
+          onPlusClick={() =>
+            this.panZoomContainer && this.panZoomContainer.zoomIn()
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/PipelineViz.jsx]: You do no need `() =>` if you are not going to pass any parameters.

File: app/assets/src/components/views/pipeline_viz.scss [hunk 10]
```diff
@@ -28,14 +30,15 @@
...
+  justify-content: space-between;
+  align-items: center;
+  flex-wrap: wrap;
 }
 
-.closeButton {
+.closeIcon {
   padding: 3px 12px;
-  cursor: pointer;
+  overflow: visible !important;
```
Comment (tfrcarvalho) [inline app/assets/src/components/views/pipeline_viz.scss]: Why do you need important?

File: app/assets/src/components/views/PipelineViz.jsx (hunk 5 above)
Comment (tfrcarvalho) [reply app/assets/src/components/views/PipelineViz.jsx]: Can you just make it 100% transparent?

File: app/assets/src/components/views/pipeline_viz.scss (hunk 10 above)
Comment (emilyzhong) [reply app/assets/src/components/views/pipeline_viz.scss]: The icon wasn't showing, but it was due to the padding — resolved!

File: app/assets/src/components/views/PipelineViz.jsx (hunk 5 above)
Comment (emilyzhong) [reply app/assets/src/components/views/PipelineViz.jsx]: Done — didn't realize edges had a `color.opacity` field

**Pull request merged**